addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerHandler.java

```diff
@@ -0,0 +1,172 @@
+ */
+public class UniFiControllerHandler extends BaseBridgeHandler implements Runnable {
+
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections
+            .singleton(UniFiBindingConstants.THING_TYPE_CONTROLLER);
+
+    private Logger logger = LoggerFactory.getLogger(UniFiControllerHandler.class);
+
+    private UniFiControllerConfig config;
+
+    private volatile UniFiController controller; /* mgb: volatile because accessed from multiple threads */
+
+    private ScheduledFuture<?> refreshJob;
+
+    public UniFiControllerHandler(Bridge bridge) {
+        super(bridge);
+    }
+
+    protected UniFiController getController() {
+        return controller;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // nop - read-only binding
+        logger.info("Ignoring command = {} for channel = {} - the UniFi binding is read-only!", command, channelUID);
+    }
+
+    @Override
+    public void initialize() {
+        Version version = bundleContext.getBundle().getVersion();
+        logger.info("UniFi Binding v{}", version);
+
+        // mgb: called when the config changes
+
+        cancelRefreshJob();
+
+        config = getConfig().as(UniFiControllerConfig.class);
+        logger.debug("Initializing the UniFi Controller Handler with config = {}", config);
+
+        try {
+            controller = new UniFiController(config);
+        } catch (UniFiException e) {
+            logger.error("Error configuring the UniFi Controller: {}", e.getMessage());
+            updateStatus(OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    e.getMessage() + " - Please double-check your configuration and try again.");
+            controller = null;
+            return;
+        }
+
+        scheduleRefreshJob();
+    }
+
+    @Override
+    public void dispose() {
+        cancelRefreshJob();
+    }
+
+    private void scheduleRefreshJob() {
+        logger.debug("Scheduling refresh job every {}s", config.getRefresh().longValue());
+        refreshJob = scheduler.scheduleWithFixedDelay(this, 0, config.getRefresh().longValue(), TimeUnit.SECONDS);
+    }
+
+    private void cancelRefreshJob() {
+        if (refreshJob != null) {
+            logger.debug("Cancelling refresh job");
+            refreshJob.cancel(true);
+            refreshJob = null;
+        }
+    }
+
+    private void refresh() {
+        ThingStatus status = getThing().getStatus();
+        ThingStatusDetail statusDetail = getThing().getStatusInfo().getStatusDetail();
+        String statusDescription = getThing().getStatusInfo().getDescription();
+
+        if (controller == null) {
+            // mgb: this should never happen
+            logger.warn("UniFi Controller is NULL - this should never happen!");
+        } else {
+            // mgb: we're offline, we need to try to login
+            if (status == INITIALIZING || (status == OFFLINE && statusDetail == COMMUNICATION_ERROR)) {
```

Comment:
@kaikreuzer I never saw such logic before in a binding is this expected for you?

Reply:
After looking over this again and again, I kept saying to myself "what were you thinking here?!?". It finally dawned on me why I did it this way.

I was originally checking whether the controller credentials were valid in the `initialize()` method, but it occurred to me the controller could be offline / unreachable and I would wrongly report this as an invalid configuration (invalid credentials).

I'm going to think about how to clean this up to make it clearer.